pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,351 @@
+          .collect(Collectors.toSet());
+      if (segmentsForOfflineTable.isEmpty()) {
+        LOGGER.warn("Skip generating task: {} for table: {}, no segment is found to merge.", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      // From lowest to highest granularity
+      String lowerGranularity = null;
+      long lowerGranularityWatermarkMs = -1;
+      for (Map.Entry<String, MergeProperties> entry : mergeProperties.entrySet()
+          .stream()
+          .sorted((e1, e2) -> Long.compare(TimeUtils.convertPeriodToMillis(e1.getValue().getBucketTimePeriod()),
+              TimeUtils.convertPeriodToMillis(e2.getValue().getBucketTimePeriod())))
+          .collect(Collectors.toList())) {
+        String granularity = entry.getKey();
+        // Only schedule 1 task per granularity
+        if (inCompleteGranularities.containsKey(granularity)) {
+          LOGGER.info("Found incomplete tasks: {} for same granularity: {} and table: {}. Skipping task generation.",
+              inCompleteGranularities.get(granularity), granularity, offlineTableName);
+          continue;
+        }
+
+        MergeProperties mergeProperty = entry.getValue();
+        Set<OfflineSegmentZKMetadata> candidateSegments = new HashSet<>(segmentsForOfflineTable);
+
+        // Get the bucket size and buffer
+        long bucketMs = TimeUtils.convertPeriodToMillis(mergeProperty.getBucketTimePeriod());
+        long bufferMs = TimeUtils.convertPeriodToMillis(mergeProperty.getBufferTimePeriod());
+
+        // Get watermark from MergeRollupTaskMetadata ZNode. WindowStart = watermark. WindowEnd = windowStart + bucket.
+        long windowStartMs = getWatermarkMs(offlineTableName, candidateSegments, bucketMs, granularity);
+        long windowEndMs = windowStartMs + bucketMs;
+
+        // Check that execution window endTimeMs <= now - bufferTime
+        if (windowEndMs > System.currentTimeMillis() - bufferMs) {
+          LOGGER.info(
+              "Window with start: {} and end: {} is not older than buffer time: {} configured as {} ago. Skipping task generation: {}",
+              windowStartMs, windowEndMs, bufferMs, bufferMs, taskType);
+          lowerGranularity = granularity;
+          lowerGranularityWatermarkMs = windowStartMs;
+          continue;
+        }
+
+        // Check that execution window endTimeMs <= waterMark of the lower granularity
+        if (lowerGranularity != null && windowEndMs > lowerGranularityWatermarkMs) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of granularity: {} is not older than the watermark: {} of granularity: (). Skipping task generation: {}",
+              windowStartMs, windowEndMs, granularity, lowerGranularityWatermarkMs, lowerGranularity, taskType);
+          lowerGranularity = granularity;
+          lowerGranularityWatermarkMs = windowStartMs;
+          continue;
+        }
+
+        // Filter out segments that are already merged for the granularity
```

Comment:
Where do we update the segment zk metadata?

Reply:
The segment zk metadata is updated when uploading the merged segment.